File: src/main/java/com/back/mozu/domain/reservation/service/ReservationAsyncProcessor.java
package com.back.mozu.domain.reservation.service;
import com.back.mozu.domain.queue.service.LockService;  // 추가
import com.back.mozu.domain.reservation.entity.Reservation;
import com.back.mozu.domain.reservation.entity.TimeSlot;
import com.back.mozu.domain.reservation.repository.ReservationRepository;
import com.back.mozu.domain.reservation.repository.TimeSlotRepository;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import org.springframework.orm.ObjectOptimisticLockingFailureException;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;
import java.util.UUID;
// 재고 관리
@Service
@RequiredArgsConstructor
public class ReservationAsyncProcessor {
    private final ReservationRepository reservationRepository;
    private final TimeSlotRepository timeSlotRepository;
    private final LockService lockService;  // 추가
    // 새 버전 - 예약 처리
    @Async
    @Transactional
    public void processReservation(UUID reservationId, UUID timeSlotId, int guestCount) {
        // 1. 예약 시도 기록과 타임슬롯을 조회
        Reservation reservation = reservationRepository.findById(reservationId)
                .orElseThrow(() -> new IllegalArgumentException("예약 기록을 찾을 수 없습니다."));
        TimeSlot timeSlot = timeSlotRepository.findById(timeSlotId)
                .orElseThrow(() -> new IllegalArgumentException("타임슬롯을 찾을 수 없습니다."));
        // 분산 락 토큰 (이 요청 고유 식별자)  // 추가
        String lockToken = reservationId.toString();  // 추가
        boolean lockAcquired = false;  // 추가
        try {
            // 2. 분산 락 획득 시도  // 추가
            lockAcquired = lockService.acquireLock(timeSlotId.toString(), lockToken);  // 추가
            if (!lockAcquired) {  // 추가
                // 락 획득 실패 → 다른 요청이 처리 중  // 추가
                reservation.cancelReservation("LOCK_ACQUIRE_FAIL");  // 추가
                return;  // 추가
            }  // 추가
            // 3. 재고 점유 (낙관적 락 발동)
            // occupy 내부에서 stock < guestCount 체크와 차감을 동시에 함
            timeSlot.occupy(guestCount);
            // 4. 예약 상태 변경 (PENDING -> CONFIRMED)
            reservation.confirmReservation();
        } catch (ObjectOptimisticLockingFailureException e) {
            // [CASE A] 다른 유저가 찰나의 순간에 먼저 가져감 (낙관적 락 충돌)
            // 재고는 실제로 안 깎였으므로 예약만 취소 처리
            reservation.cancelReservation("OPTIMISTIC_LOCK_FAIL");
        } catch (IllegalArgumentException | IllegalStateException e) {
            // [CASE B] 재고 부족이거나 이미 취소된 건인 경우
            reservation.cancelReservation("RESERVATION_FAILED");
            // 혹시 모르니 여기서도 재고는 건드리지 않음 (이미 occupy에서 실패했으므로)
        } catch (Exception e) {
            // [CASE C] 그 외 예상치 못한 시스템 에러 (DB 연결 끊김 등)
            reservation.cancelReservation("SYSTEM_ERROR");
            // 만약 occupy는 성공했는데 여기서 터졌다면? 안전하게 재고를 돌려줌
            timeSlot.release(guestCount);
        } finally {
            // 5. 락 반드시 해제 (성공/실패 무관, 데드락 방지)  // 추가
            if (lockAcquired) {  // 추가
                lockService.releaseLock(timeSlotId.toString(), lockToken);  // 추가
            }  // 추가
        }
    }
}

File: src/main/java/com/back/mozu/domain/reservation/service/ReservationAsyncProcessor.kt
package com.back.mozu.domain.reservation.service

import com.back.mozu.domain.reservation.entity.ReservationStatus
import com.back.mozu.domain.reservation.repository.ReservationRepository
import com.back.mozu.domain.reservation.repository.TimeSlotRepository
import org.redisson.api.RedissonClient
import org.slf4j.LoggerFactory
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.orm.ObjectOptimisticLockingFailureException
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.transaction.support.TransactionSynchronization
import org.springframework.transaction.support.TransactionSynchronizationManager
import java.util.UUID
import java.util.concurrent.TimeUnit

@Service
class ReservationAsyncProcessor(
    private val reservationRepository: ReservationRepository,
    private val timeSlotRepository: TimeSlotRepository,
    private val redissonClient: RedissonClient,
    private val redisTemplate: RedisTemplate<String, String>,
    private val reservationStatusService: ReservationStatusService
) {

    @Async
    @Transactional(noRollbackFor = [IllegalStateException::class])
    fun processReservation(reservationId: UUID, timeSlotId: UUID, guestCount: Int) {
        val timeSlotIdStr = timeSlotId.toString()
        val occupiedKey = "isOccupied:$timeSlotIdStr"
        val lockKey = "lock:timeslot:$timeSlotIdStr"

        val isOccupied = redisTemplate.opsForValue().get(occupiedKey)?.toBoolean() ?: false
        if (isOccupied) {
            reservationStatusService.cancelReservationSafely(reservationId, "ALREADY_OCCUPIED_FAIL_FAST")
            return
        }

        var lockAcquired = false
        var stockOccupied = false
        val lock = redissonClient.getLock(lockKey)

        try {
            lockAcquired = lock.tryLock(5, -1, TimeUnit.SECONDS)

            if (!lockAcquired) {
                reservationStatusService.cancelReservationSafely(reservationId, "LOCK_ACQUIRE_FAIL")
                return
            }

            val reservation = reservationRepository.findByIdWithLock(reservationId)
                ?: throw IllegalArgumentException("예약 기록을 찾을 수 없습니다.")

            if (reservation.status != ReservationStatus.PENDING) {
                reservationStatusService.cancelReservationSafely(reservationId, "ALREADY_PROCESSED")
                return
            }

            val timeSlot = timeSlotRepository.findByIdWithLock(timeSlotId)
                ?: throw IllegalArgumentException("타임슬롯을 찾을 수 없습니다.")

            timeSlot.occupy(guestCount)
            stockOccupied = true

            reservation.confirmReservation()

            if (timeSlot.stock == 0) {
                redisTemplate.opsForValue().set(occupiedKey, "true")
            }

        } catch (e: ObjectOptimisticLockingFailureException) {
            reservationStatusService.cancelReservationSafely(reservationId, "OPTIMISTIC_LOCK_FAIL")

        } catch (e: IllegalArgumentException) {
            reservationStatusService.cancelReservationSafely(reservationId, "RESERVATION_FAILED")

        } catch (e: IllegalStateException) {
            if (stockOccupied) {
                timeSlotRepository.findByIdWithLock(timeSlotId)?.release(guestCount)
            }
            reservationStatusService.cancelReservationSafely(reservationId, "RESERVATION_FAILED")
            redisTemplate.delete(occupiedKey)

        } catch (e: Exception) {
            log.error(
                "비동기 예약 처리 중 시스템 예외 발생: reservationId={}, timeSlotId={}, guestCount={}",
                reservationId, timeSlotId, guestCount, e,
            )
            if (stockOccupied) {
                timeSlotRepository.findByIdWithLock(timeSlotId)?.release(guestCount)
            }
            reservationStatusService.cancelReservationSafely(reservationId, "SYSTEM_ERROR")
            redisTemplate.delete(occupiedKey)

        } finally {
            if (lockAcquired) {
                if (TransactionSynchronizationManager.isSynchronizationActive()) {
                    TransactionSynchronizationManager.registerSynchronization(
                        object : TransactionSynchronization {
                            override fun afterCompletion(status: Int) {
                                if (lock.isLocked && lock.isHeldByCurrentThread) {
                                    lock.unlock()
                                    log.debug("Transaction 커밋 후 Redisson 락 해제 완료: {}", lockKey)
                                }
                            }
                        }
                    )
                } else {
                    if (lock.isLocked && lock.isHeldByCurrentThread) {
                        lock.unlock()
                    }
                }
            }
        }
    }

    companion object {
        private val log = LoggerFactory.getLogger(ReservationAsyncProcessor::class.java)
    }
}